### PR TITLE
Added Components for PresentationML

### PR DESCRIPTION
**Pull request by kobsy**
This is a major expansion and reworking of DrawingML, taking advantage of expanded ways of specifying properties in the newer version of `openxml-package`.

Some elements are obviated by this use of properties and will eventually be able to be deleted. However, this will be a breaking change for gems that rely on those elements _as_ elements, and so I left that change for another day.

### CODE REVIEW

File: lib/openxml/drawingml/elements/paragraph.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        COORDINATE_REGEX = /^(-?[0-9]+(:?\.[0-9]+)?)(:?mm|cm|in|pt|pc|pi)?$/
+
+        coordinate = ->(value) { raise ArgumentError unless value.to_s =~ coordinate_regex }
+        indent_range = lambda { |value|
+          matched = value.to_s =~ COORDINATE_REGEX
+          raise ArgumentError unless matched || (-51_206_400..51_206_400).cover?(value)
+        }
+        text_margin = lambda { |value|
+          matched = value.to_s =~ COORDINATE_REGEX
+          raise ArgumentError unless matched || (0..51_206_400).cover?(value)
```
Comment (boblail) [inline lib/openxml/drawingml/elements/paragraph.rb]: Should these `ArgumentError`s have a message?

File: lib/openxml/drawingml/elements/paragraph.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+          property :autonumbered_bullets, as: :bullet_style_auto_numbered
+          property :character_bullets, as: :bullet_style_character
+          property :blip_bullets, as: :bullet_style_blip
+        end
+
+        property :tab_list
+        property :default_run_properties
+        property :extension_list
+
+      private
```
Comment (boblail) [inline lib/openxml/drawingml/elements/paragraph.rb]: Is this `private` and the whitespace below it here for any reason?

File: lib/openxml/drawingml/properties/blip.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+module OpenXml
+  module DrawingML
+    module Properties
+      class Blip < OpenXml::Properties::ComplexProperty
+        include HasChildren
+        include ContainsProperties
```
Comment (boblail) [inline lib/openxml/drawingml/properties/blip.rb]: I'm OK with this style, but did you know you can also comma-separate included modules?

```ruby
include HasChildren, ContainsProperties
```

File: lib/openxml/drawingml/properties/blip.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+          attribute :link, expects: :string
+        end
+
+        # TODO: All the various blip child elements as properties
+
+        property :extension_list
+
+        def render?
+          true
+        end
```
Comment (boblail) [inline lib/openxml/drawingml/properties/blip.rb:22]: ❔ What does this do?

File: lib/openxml/drawingml/properties/body_properties.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        property :extension_list
+
+      private
+
+        def coordinate(value)
+          raise ArgumentError unless value.to_s =~ /^-?[0-9]+(\.[0-9]+)?(mm|cm|in|pt|pc|pi)?$/
+        end
+
+        def positive_coordinate(value)
+          raise ArgumentError unless value.to_s =~ /^[0-9]+(\.[0-9]+)?(mm|cm|in|pt|pc|pi)?$/
```
Comment (boblail) [inline lib/openxml/drawingml/properties/body_properties.rb]: 👆 Are these copy-and-pastes for your `COORDINATE_REGEX` from `lib/openxml/drawingml/elements/paragraph.rb`? If so, can you extract them to a common location — so the regex is defined just once?

File: lib/openxml/drawingml/properties/bullet_size_percent.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+module OpenXml
+  module DrawingML
+    module Properties
+      class BulletSizePercent < ValueProperty
+        namespace :a
+        tag :buSzPct
+
+        def valid?
+          # Regexp taken from OpenXml Spec Part 1, section 20.1.10.61
+          /0*((2[5-9])|([3-9][0-9])|([1-3][0-9][0-9])|400)%/ =~ value
```
Comment (boblail) [inline lib/openxml/drawingml/properties/bullet_size_percent.rb]: 👍 Awesome comment!

File: lib/openxml/drawingml/properties/color_scheme.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        property :dark_two, as: :color_scheme_dark_two                      # Required
+        property :light_two, as: :color_scheme_light_two                    # Required
+        property :accent_one, as: :color_scheme_accent_one                  # Required
+        property :accent_two, as: :color_scheme_accent_two                  # Required
+        property :accent_three, as: :color_scheme_accent_three              # Required
+        property :accent_four, as: :color_scheme_accent_four                # Required
+        property :accent_five, as: :color_scheme_accent_five                # Required
+        property :accent_six, as: :color_scheme_accent_six                  # Required
+        property :hyperlink, as: :color_scheme_hyperlink                    # Required
+        property :followed_hyperlink, as: :color_scheme_followed_hyperlink  # Required
```
Comment (boblail) [inline lib/openxml/drawingml/properties/color_scheme.rb]: What does the `# Required` comment mean in this file? — didn't you add a `required: true` option to `property` (or was that `attribute`)?

File: lib/openxml/drawingml/properties/color_xform_alpha.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+module OpenXml
+  module DrawingML
+    module Properties
+      class ColorXformAlpha < ValueProperty
+        namespace :a
+        tag :alpha
+
+        def valid?
+          /^((100)|([0-9][0-9]?))(\.[0-9][0-9]?)?%$/ =~ value
```
Comment (boblail) [inline lib/openxml/drawingml/properties/color_xform_alpha.rb]: In **lib/openxml/drawingml/properties/color_xform_alpha_offset.rb**, You labeled `/^-?((100)|([0-9][0-9]?))(\.[0-9][0-9]?)?%$/ =~ value` as _ST_FixedPercentage, Part 1 - 20.1.10.24_. Is there no such label for this one?

This "positive percentage" regex is repeated in **color_xform_tint.rb** and **color_xform_shade.rb** too.

What you extracted these regexes to constants? Like:

```ruby
module OpenXml
  module DrawingML
     ST_FixedPercentage = `/^-?((100)|([0-9][0-9]?))(\.[0-9][0-9]?)?%$/` # §20.1.10.24
  end
end
```

If it's in the DrawingML namespace, it wouldn't clobber regexes defined in other openxml gems; but it should be available to Elements and Properties without you having to give its fully-qualified name.

Would that be a way of naming all these special regexes?

File: lib/openxml/drawingml/properties/custom_geometry.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+      class CustomGeometry < SimplePropertyContainerProperty
+        namespace :a
+        tag :custGeom
+
+        property :shape_adjustment_list
+        property :geometry_guide_list
+        # property :adjustable_handle_list # ahLst
+        # property :connection_list # cxnLst
+        # property :rectangle # rect
+        property :path_list # Required!
```
Comment (boblail) [inline lib/openxml/drawingml/properties/custom_geometry.rb]: What's with commented-out properties?

The `# Required!` comment — again, did you implement that?

File: lib/openxml/drawingml/properties/color_xform_alpha.rb (hunk 8 above)
Comment (boblail) [reply lib/openxml/drawingml/properties/color_xform_alpha.rb]: Look like you name it in **dash_stop.rb** and **gradient_stop.rb** and **line_join_mitre.rb**

File: lib/openxml/drawingml/properties/path.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
+# require "openxml/drawingml/properties/path_close"
+# require "openxml/drawingml/properties/path_move_to"
+# require "openxml/drawingml/properties/path_line_to"
+# require "openxml/drawingml/properties/path_arc_to"
+# require "openxml/drawingml/properties/path_quadratic_bezier_to"
+# require "openxml/drawingml/properties/path_cubic_bezier_to"
```
Comment (boblail) [inline lib/openxml/drawingml/properties/path.rb:6]: What are the commented-out requires for?

File: lib/openxml/drawingml/properties/path.rb [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+module OpenXml
+  module DrawingML
+    module Properties
+      class Path < OpenXml::Properties::ContainerProperty
+        include OpenXml::HasAttributes
+        namespace :a
+        tag :path
+        # TODO: Implement child classes
+        # child_classes :path_close, :path_move_to, :path_line_to, :path_arc_to,
+        #               :path_quadradic_bezier_to, :cubic_bezier_to
```
Comment (boblail) [inline lib/openxml/drawingml/properties/path.rb:17]: commented code

**Review by boblail — CHANGES_REQUESTED** [review]
This is tremendous! The DSL for defining and testing these OpenXml elements is good and keeps getting better. The tests are good. All of this is a prodigious amount of code! 😆 — and so much of it is neatly declarative. 👍 

I think the main thing that could be improved is the scattering of magic constants throughout the files. There are a lot of copies of regular expressions for matching coordinate-numbers-with-units and percentages. These special bits of knowledge should be kept in one place so that changes to them are less likely to introduce bugs of inconsistency. Already, there's inconsistency in how they're documented: sometimes they're named, sometimes they're annotated with a comment, sometimes not.

The other thing I would suggest **(less for _this_ PR and more for going forward)** is to use comments that "link" back to the spec more liberally. You've got a few — that indicate where a regex is defined in the spec — and I think those are handy references. 

Lastly, do you think — if we ever refer to this commit in the future — that its description is helpful? I can't see when the list of commits which were squashed into this one is ever information that future-us would need to know.

File: lib/openxml/drawingml/properties/color_scheme.rb (hunk 7 above)
Comment (kobsy) [reply lib/openxml/drawingml/properties/color_scheme.rb]: That was one of the ideas I was playing around with, but never implemented. In the meantime, I left the comments here to remind me or whoever might go back and implement such a thing down the line that these properties should receive the `required: true` (or equivalent option)

File: lib/openxml/drawingml/properties/color_xform_alpha.rb (hunk 8 above)
Comment (kobsy) [reply lib/openxml/drawingml/properties/color_xform_alpha.rb]: ☝️ I like this idea of extracting them into constants!

File: lib/openxml/drawingml/properties/path.rb (hunk 10 above)
Comment (kobsy) [reply lib/openxml/drawingml/properties/path.rb:6]: These child classes are not implemented yet (and for our immediate purposes, didn't need to be). I liked keeping the requires and the actual `child_classes` declaration (below), though, as a reference for implementing these in the future.

File: lib/openxml/drawingml/properties/blip.rb (hunk 4 above)
Comment (kobsy) [reply lib/openxml/drawingml/properties/blip.rb:22]: By default, a property is only rendered if it a) has attributes defined or b) has children of some form. There are a handful of exceptions (where you'd want a blank tag to render anyways), and this is one of them. It forces the render by overriding the `render?` method.

File: lib/openxml/drawingml/properties/blip.rb (hunk 3 above)
Comment (kobsy) [reply lib/openxml/drawingml/properties/blip.rb]: 👍

File: lib/openxml/drawingml/elements/paragraph.rb (hunk 2 above)
Comment (kobsy) [reply lib/openxml/drawingml/elements/paragraph.rb]: I think this was copy/pasted from the element I moved over from Pptx

File: lib/openxml/drawingml/elements/paragraph.rb (hunk 1 above)
Comment (kobsy) [reply lib/openxml/drawingml/elements/paragraph.rb]: Yeah, probably so...

File: lib/openxml/drawingml/properties/color_scheme.rb (hunk 7 above)
Comment (boblail) [reply lib/openxml/drawingml/properties/color_scheme.rb]: What would happen if you just passed `required: true` to each of these properties _and it just didn't do anything_?

Would the `property` method raise an exception b/c it received an unknown option?

It'd be just like what you're doing, documenting the methods, but (obviously) in a way that we could work with programmatically — or just implement some day.

Comment (kobsy) [reply lib/openxml/drawingml/properties/color_scheme.rb]: Good question! I'll give it a try and see if I can change these comments into code.

File: lib/openxml/drawingml/properties/path.rb (hunk 10 above)
Comment (boblail) [reply lib/openxml/drawingml/properties/path.rb:6]: 👌

File: lib/openxml/drawingml/properties/blip.rb (hunk 4 above)
Comment (boblail) [reply lib/openxml/drawingml/properties/blip.rb:22]: Ah! Ok. Thanks!

**Comment by kobsy** [comment]
Ready for re-review, @boblail, with the caveat that I'll need to update the `Gemfile` once we cut a new version of `openxml-package` with the `required` placeholders.

File: Gemfile [hunk 12]
```diff
@@ -3,4 +3,4 @@ source "https://rubygems.org"
 gemspec
 
 # Optional: use the development version of openxml-package
-# gem "openxml-package", github: "openxml/openxml-package", branch: "master"
+gem "openxml-package", github: "openxml/openxml-package", branch: "finish-requirements"
```
Comment (boblail) [inline Gemfile]: Now that that's merged, let's change this back.

File: lib/openxml/drawingml/elements/paragraph.rb [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+          is_integer_in_range = value.is_a?(Integer) && (-51_206_400..51_206_400).cover?(value)
+          is_measure = value.is_a?(String) && value =~ OpenXml::DrawingML::ST_UniversalMeasure
+          raise ArgumentError, message unless is_integer_in_range || is_measure
+        }
+        text_margin = lambda { |value|
+          message = "Invalid text margin: Must be an integer between 0 and 51,206,400 or include units"
+          is_integer_in_range = value.is_a?(Integer) && (0..51_206_400).cover?(value)
+          is_measure = value.is_a?(String) && value =~ OpenXml::DrawingML::ST_PositiveUniversalMeasure
+          raise ArgumentError, message unless is_integer_in_range || is_measure
+        }
```
Comment (boblail) [inline lib/openxml/drawingml/elements/paragraph.rb]: ⬇️ :art: Should these probably be `do...end` blocks versus `{...}` blocks?

File: lib/openxml/drawingml/properties/custom_geometry.rb [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+      class CustomGeometry < SimplePropertyContainerProperty
+        namespace :a
+        tag :custGeom
+
+        property :shape_adjustment_list
+        property :geometry_guide_list
+        # property :adjustable_handle_list # ahLst
+        # property :connection_list # cxnLst
+        # property :rectangle # rect
+        property :path_list, required: true
```
Comment (boblail) [inline lib/openxml/drawingml/properties/custom_geometry.rb]: Are these just to remember what else could be implemented for this property?

File: lib/openxml/drawingml/properties/default_shape_properties.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+module OpenXml
+  module DrawingML
+    module Properties
+      class DefaultShapeProperties < SimplePropertyContainerProperty
+        property :shape_properties, required: true
+        property :body_properties, required: true
+        property :list_style, required: true
+
+        # TODO: style
+        property :extension_list
```
Comment (boblail) [inline lib/openxml/drawingml/properties/default_shape_properties.rb]: What does this TODO mean?

File: lib/openxml/drawingml/properties/custom_geometry.rb (hunk 14 above)
Comment (boblail) [reply lib/openxml/drawingml/properties/custom_geometry.rb]: (In other places, I've seen you use `# TODO: add `sp3d` property` — which is clearer what's meant by it)

File: Gemfile (hunk 12 above)
Comment (kobsy) [reply Gemfile]: I'd actually love to up the dependency number in the `gemspec` file and re-comment this line out once we cut a new version of the `openxml-package` gem.

**Review by boblail — APPROVED** [review]
👍 I'm going to put a **hold** on it (not a **review-hold**) since it needs:

 1. squashing
 2. to use `v0.3.2` of `openxml-package`

The only thing I'll comment on is it looks like you use a few different patterns for commenting-out code for future-implementation. It might help to settle on one — for the openxml project — and be consistent.

File: lib/openxml/drawingml/properties/custom_geometry.rb (hunk 14 above)
Comment (kobsy) [reply lib/openxml/drawingml/properties/custom_geometry.rb]: I can add the explicit `TODO: add`, since that's exactly what this is meant to indicate.

**Comment by kobsy** [comment]
I made those style changes, made the `TODO`s more consistent, and upped our gem dependency to 0.3.2. Ready for another look, @boblail.

**Pull request merged**